File: backend/modules/petclinic/app/petclinic/api/pets/addPet/AddPetService.java
package petclinic.api.pets.addPet;

import akka.actor.ActorRef;
import akka.pattern.AskSupport;
import akka.pattern.PatternsCS;
import petclinic.api.model.OwnerId;
import petclinic.api.model.PetType;
import petclinic.model.OwnerMessages;

import javax.inject.Inject;
import javax.inject.Named;
import java.time.LocalDate;
import java.util.Optional;
import java.util.concurrent.CompletionStage;

import static petclinic.api.Constants.TIMEOUT;

public class AddPetService implements AskSupport {

    @Inject
    @Named("owner")
    private ActorRef owner;

    public CompletionStage<Optional<AddPetApiResponse>> addPet(OwnerId ownerId, String name, LocalDate birthday, PetType type) {
        return PatternsCS.ask(owner, OwnerMessages.addPet(ownerId.id(), name, birthday, type.getId(), type.getName()), TIMEOUT)
                .thenApply(obj -> (OwnerMessages.AddPetResponse) obj).thenApply(obj -> {
                    if (obj instanceof OwnerMessages.OwnerNotFound) {
                        return Optional.empty();
                    } else {
                        OwnerMessages.PetAddedResponse data = (OwnerMessages.PetAddedResponse) obj;
                        AddPetApiResponse resp = AddPetApiResponse.apply(data.id());
                        return Optional.of(resp);
                    }
                });
    }


}

File: backend/modules/petclinic/app/petclinic/api/pets/addPet/AddPetService.scala
package petclinic.api.pets.addPet

import java.time.LocalDate

import akka.actor.ActorRef
import javax.inject.{Inject, Named}
import petclinic.api.Constants
import petclinic.api.model.{OwnerId, PetType}
import petclinic.model.owners.owner.Protocol.{AddPetCommand, AddPetResponse, OwnerNotFoundPetAddedResponse, PetAddedResponse}

import scala.concurrent.{ExecutionContext, Future}

class AddPetService @Inject()(@Named(petclinic.api.Constants.OWNER_AGGREGATE_ACTOR_NAME) owner: ActorRef)(implicit ec: ExecutionContext) {

  def addPet(ownerId: OwnerId, nameIn: String, birthdayIn: LocalDate, `type`: PetType): Future[Option[AddPetApiResponse]] = {
    akka.pattern.Patterns.ask(owner, AddPetCommand(ownerId.id, nameIn, birthdayIn, `type`.getId, `type`.getName), Constants.TIMEOUT)
      .mapTo[AddPetResponse]
      .map {
        case PetAddedResponse(id, name, birthday, petTypeId) => Some(AddPetApiResponse(id))
        case OwnerNotFoundPetAddedResponse => None
      }
  }

}
